evap/grades/tests.py

```diff
@@ -144,30 +142,77 @@ def test_toggle_no_grades(self):
         self.assertEqual(response.status_code, 200)
         course = Course.objects.get(id=course.id)
         self.assertTrue(course.gets_no_grade_documents)
         # course should get published here
         self.assertEqual(course.state, "published")
         self.assertEqual(len(mail.outbox), course.num_participants + course.contributions.exclude(contributor=None).count())
 
-        response = self.app.post("/grades/toggle_no_grades", params={"course_id": course.id}, user="grade_publisher")
+        response = self.app.post("/grades/toggle_no_grades", params={"course_id": course.id}, user="grader")
         self.assertEqual(response.status_code, 200)
         course = Course.objects.get(id=course.id)
         self.assertFalse(course.gets_no_grade_documents)
 
-    def test_grade_activation(self):
-        self.activation.is_active = True
-        self.activation.save()
-
+    def test_grade_document_download_after_archiving(self):
         # upload grade document
         self.helper_upload_grades(self.course, final_grades=False)
         self.assertGreater(self.course.midterm_grade_documents.count(), 0)
 
         url = "/grades/download/" + str(self.course.midterm_grade_documents.first().id)
         self.get_assert_200(url, "student")  # grades should be downloadable
 
-        self.activation.is_active = False
-        self.activation.save()
-        self.get_assert_error(url, "student")  # grades should not be downloadable anymore
+        self.semester.archive_grade_documents()
+        self.get_assert_error(url, "student", status=404)  # grades should not be downloadable anymore
+
+
+class GradeDocumentIndexTest(ViewTest):
+    url = '/grades/'
+    test_users = ['grader']
+
+    @classmethod
+    def setUpTestData(cls):
+        mommy.make(UserProfile, username="grader", groups=[Group.objects.get(name="Grade publisher")])
+        cls.semester = mommy.make(Semester, grade_documents_are_archived=False)
+        cls.archived_semester = mommy.make(Semester, grade_documents_are_archived=True)
+
+    def test_visible_semesters(self):
+        page = self.get_assert_200(self.url, user="grader")
+        self.assertIn(self.semester.name, page)
+        self.assertNotIn(self.archived_semester.name, page)
+
+
+class GradeDocumentSemesterViewTest(ViewTest):
+    url = '/grades/semester/1'
+    test_users = ['grader']
+
+    @classmethod
+    def setUpTestData(cls):
+        mommy.make(UserProfile, username="grader", groups=[Group.objects.get(name="Grade publisher")])
+        semester = mommy.make(Semester, pk=1, grade_documents_are_archived=False)
+        mommy.make(Semester, pk=2, grade_documents_are_archived=True)
+        cls.semester_course = mommy.make(Course, semester=semester, state="prepared")
+
+    def test_semester_pages(self):
+        page = self.get_assert_200(self.url, user="grader")
+        self.assertIn(self.semester_course.name, page)
+        self.get_assert_error('/grades/semester/2', user="grader")
```

Comment:
in the first two lines we're checking that the course is actually listed on the page. should i still remove this part?
the third line then tests for an archived semester, so it could be moved to a different test.

Reply:
no you're right, checking whether the course is there makes sense. in the other test, checking for the button is really just a smoke test which i think is covered adequately by the automatic tests that simply renders the page.